### PR TITLE
Adding unit test to cover NullReferenceException when passing null to a collection

### LINKED ISSUE #37078: NullReferenceException calling JsonSerializer.ReadAsync<T>()
As per https://github.com/dotnet/corefx/issues/36611#issuecomment-485161258, I have an existing ASP.NET Core 2.2 application I've updated to use ASP.NET Core 3.0 preview 4 where I tried to plug in System.Text.Json into Refit's `IContentSerializer` interface as a replacement for Newtonsoft.Json.

However, when trying to deserialize a JSON stream for various application models to their POCOs, a `NullReferenceException` is thrown.

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at void System.Text.Json.Serialization.ReadStackFrame.SetReturnValue(object value, JsonSerializerOptions options, ref ReadStackFrame current, bool setPropertyDirectly)
   at bool System.Text.Json.Serialization.JsonSerializer.HandleNull(ref Utf8JsonReader reader, ref ReadStack state, JsonSerializerOptions options)
   at void System.Text.Json.Serialization.JsonSerializer.ReadCore(JsonSerializerOptions options, ref Utf8JsonReader reader, ref ReadStack state)
   at void System.Text.Json.Serialization.JsonSerializer.ReadCore(ref JsonReaderState readerState, bool isFinalBlock, Span<byte> buffer, JsonSerializerOptions options, ref ReadStack state)
   at async ValueTask<TValue> System.Text.Json.Serialization.JsonSerializer.ReadAsync<TValue>(Stream utf8Json, Type returnType, JsonSerializerOptions options, CancellationToken cancellationToken)
   at TResult System.Threading.Tasks.ValueTask<TResult>.get_Result()
```

A repro for the issue can be found here: https://github.com/martincostello/JsonSerializer-NullReferenceException-Repro

The object graph is probably more verbose that needed for a minimum LoC repro, but except for the models and properties being renamed and the JSON string simplified, this is the same as the application that presented the issue.

The essence of the code that throws is:

```cs
var json = "{ ... large JSON model with nested objects and null values ... }";
var buffer = Encoding.UTF8.GetBytes(json);
var utf8Json = new MemoryStream(buffer);
var options = new JsonSerializerOptions();

var model = await JsonSerializer.ReadAsync<RootObject>(utf8Json, options);
```

/cc @BrennanConroy

### PR DESCRIPTION
**Pull request by Jozkee**
Fixes #37078 
Problem: in v3.0.0-preview4, passing `null` to a collection through `JsonSerializer.ReadAsync` will throw an unexpected NullRef error.

Fix: Since I was able to reproduce the reported issue in v3.0.0-preview4 with the unit test specified in #37078, I added a simplified version of it to our tests. This error does not happen in our latest build and therefore is considered as corrected.

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs [hunk 1]
```diff
@@ -79,5 +81,17 @@ public static async Task ReadPrimitivesWithTrailingTriviaAsync()
                 Assert.Equal(1, i);
             }
         }
+
+        [Fact]
+        public static async Task ReadCollectionPassingNullValue()
+        {
+            using (MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes("null")))
+            {
+                JsonSerializerOptions options = new JsonSerializerOptions();
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: nit: This is not necessary since the default options will be used anyway if you don't pass anything to the `JsonSerializer`:
https://github.com/dotnet/corefx/blob/23faf5d032feef556f5399b46a92a6e4aef7d84a/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L19

So, just passing the stream (without options) should be sufficient.

File: src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs [hunk 2]
```diff
@@ -79,5 +81,17 @@ public static async Task ReadPrimitivesWithTrailingTriviaAsync()
...
+
+                var collection  = await JsonSerializer.ReadAsync<IList<object>>(stream, options);
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: nit: Don't use `var` here. Generally, we only use `var` when it can be statically inferred from the right hand side what the type was (for instance if the `new` keyword was used or an explicit cast was there making it obvious). It helps reviewers/reading the code.

https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md#c-coding-style

```suggestion
                IList<object> collection  = await JsonSerializer.ReadAsync<IList<object>>(stream, options);
```

File: src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs [hunk 3]
```diff
@@ -2,6 +2,8 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections;
+using System.Collections.Generic;
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: nit: Do we need both these using directives? As far as I can tell, we only added `IList<T>`, so the generic one should be enough.

File: src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs [hunk 4]
```diff
@@ -79,5 +81,17 @@ public static async Task ReadPrimitivesWithTrailingTriviaAsync()
                 Assert.Equal(1, i);
             }
         }
+
+        [Fact]
+        public static async Task ReadCollectionPassingNullValue()
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: nit: I know we aren't consistent (at least in tests), but generally async methods have `Async` in the method name, as a suffix.

```suggestion
        public static async Task ReadCollectionPassingNullValueAsync()
```

File: src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs [hunk 5]
```diff
@@ -79,5 +81,17 @@ public static async Task ReadPrimitivesWithTrailingTriviaAsync()
...
+        {
+            using (MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes("null")))
+            {
+                JsonSerializerOptions options = new JsonSerializerOptions();
+
+                var collection  = await JsonSerializer.ReadAsync<IList<object>>(stream, options);
+                Assert.Null(collection);
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: nit: Can you also add an attempt to read into a collection of a non-reference type in this test method, like `IList<int>`?

```C#
IList<int> anotherCollection = await JsonSerializer.ReadAsync<IList<int>>(stream);
Assert.Null(anotherCollection);
```

File: src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs (hunk 3 above)
Comment (Jozkee) [reply src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: I removed `System.Collections` directive and only left there `System.Collections.Generic`, thanks for the suggestion.

File: src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs [hunk 6]
```diff
@@ -79,5 +80,21 @@ public static async Task ReadPrimitivesWithTrailingTriviaAsync()
...
+        [Fact]
+        public static async Task ReadCollectionPassingNullValueAsync()
+        {
+            using (MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes("null")))
+            {
+                IList<object> referenceTypeCollection  = await JsonSerializer.ReadAsync<IList<object>>(stream);
+                Assert.Null(referenceTypeCollection);
+            }
+
+            using (MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes("null")))
```
Comment (danmosemsft) [inline src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: Your choice, but usually when a test has two near-identical tests inside it, it's better off being two tests. That way if it fails it's obvious which one failed. Also, if for some reason one of them starts to fail and needs to be disabled, we don't need to disable both.

Comment (ahsonkhan) [reply src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: An alternative approach could be to use `Theory` here and test the method that accepts the returning type as an argument. I would be fine with either approaches. If the intent is to actually test the generic overload (and substituting it with the non-generic API isn't feasible), then two separate tests works just as well.

```C#
[Theory]
[InlineData(typeof(IList<object>))]
[InlineData(typeof(IList<int>))]
public static async Task ReadSimpleObjectAsync(Type type)
{
    using (MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes("null")))
    {
        object collection = await JsonSerializer.ReadAsync(stream, type);
        Assert.Null(collection);
    }
}
```

Comment (Jozkee) [reply src/System.Text.Json/tests/Serialization/Stream.ReadTests.cs]: I will stick to create separate unit test methods for each scenario.

The reason is, the origin of this issue was using the generic type overload and also there are no significant differences between both signatures; as you can see, the two public methods end up calling the private overload:
https://github.com/dotnet/corefx/blob/eaa0079941923b4f903344130315f4cd332b84fe/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs#L75

**Comment by ahsonkhan** [comment]
Unrelated System.Net.Http.Functional.Tests failures:
 System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http2/GoAwayFrame_AbortAllPendingStreams_StreamFailWithExpectedException
https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F38566~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190619.80/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http2~2FGoAwayFrame_AbortAllPendingStreams_StreamFailWithExpectedException

System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http2/InboundWindowSize_Exceeded_Throw
https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F38566~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190619.80/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.SocketsHttpHandlerTest_Http2~2FInboundWindowSize_Exceeded_Throw

https://github.com/dotnet/corefx/issues/38708
